service/history/shardController.go

```diff
@@ -322,24 +322,31 @@ func (c *shardController) acquireShards() {
 	sw := c.metricsClient.StartTimer(metrics.HistoryShardControllerScope, metrics.AcquireShardsLatency)
 	defer sw.Stop()
 
-AcquireLoop:
-	for shardID := 0; shardID < c.config.NumberOfShards; shardID++ {
-		info, err := c.hServiceResolver.Lookup(string(shardID))
-		if err != nil {
-			c.logger.Error("Error looking up host for shardID", tag.Error(err), tag.OperationFailed, tag.ShardID(shardID))
-			continue AcquireLoop
-		}
+	concurrency := c.config.AcquireShardConcurrency()
+	sem := make(chan bool, concurrency)
 
-		if info.Identity() == c.host.Identity() {
-			_, err1 := c.getEngineForShard(shardID)
-			if err1 != nil {
-				c.metricsClient.IncCounter(metrics.HistoryShardControllerScope, metrics.GetEngineForShardErrorCounter)
-				c.logger.Error("Unable to create history shard engine", tag.Error(err1), tag.OperationFailed, tag.ShardID(shardID))
-				continue AcquireLoop
+	for shardID := 0; shardID < c.config.NumberOfShards; shardID++ {
+		sem <- true
+		go func(shardID int) {
```

Comment:
consider using the [worker-pool](https://gobyexample.com/worker-pools) pattern here (you can ignore the results chan in that example - which is not needed here). I think code will be more straightforward and also efficient (doesn't create goroutine per shardID) with that pattern. In addition, you can use waitGroups to wait for all goroutines to finish, which go-readers are used to seeing.

Reply:
It is slightly more verbose, but arguably easier to read. Updated.